src/NVMLDevicePool.cpp

```diff
@@ -0,0 +1,465 @@
+        }
+        else {
+            throw Exception("NVMLDevicePool::" + std::string(__func__) +
+                            ": NVML failed to get SM Frequency for accelerator " +
+                            std::to_string(accel_idx) + ".  Error: " + nvmlErrorString(nvml_result),
+                            GEOPM_ERROR_INVALID, __FILE__, __LINE__);
+        }
+        return result;
+    }
+
+    double NVMLDevicePoolImp::utilization(int accel_idx) const
+    {
+        check_accel_range(accel_idx);
+        double result = NAN;
+        nvmlUtilization_t util;
+        nvmlReturn_t nvml_result;
+
+        nvml_result =  nvmlDeviceGetUtilizationRates(m_nvml_device.at(accel_idx), &util);
+        if (nvml_result == NVML_SUCCESS) {
+            result = (double) util.gpu;
+        }
+        else {
+            throw Exception("NVMLDevicePool::" + std::string(__func__) +
+                            ": NVML failed to get GPU Utilization for accelerator " +
+                            std::to_string(accel_idx) + ".  Error: " + nvmlErrorString(nvml_result),
+                            GEOPM_ERROR_INVALID, __FILE__, __LINE__);
+        }
+        return result;
+    }
+
+    uint64_t NVMLDevicePoolImp::power(int accel_idx) const
+    {
+        check_accel_range(accel_idx);
+        uint64_t result = 0;
+        unsigned int temp;
+        nvmlReturn_t nvml_result;
+
+        nvml_result = nvmlDeviceGetPowerUsage(m_nvml_device.at(accel_idx), &temp);
+        if (nvml_result == NVML_SUCCESS) {
+            result = (uint64_t)temp;
+        }
+        else {
+            throw Exception("NVMLDevicePool::" + std::string(__func__) +
+                            ": NVML failed to get power for accelerator " +
+                            std::to_string(accel_idx) + ".  Error: " + nvmlErrorString(nvml_result),
+                            GEOPM_ERROR_INVALID, __FILE__, __LINE__);
+        }
+
+        return result;
+    }
+
+    uint64_t NVMLDevicePoolImp::frequency_status_mem(int accel_idx) const
+    {
+        check_accel_range(accel_idx);
+        uint64_t result = 0;
+        unsigned int temp;
+        nvmlReturn_t nvml_result;
+
+        nvml_result = nvmlDeviceGetClock(m_nvml_device.at(accel_idx), NVML_CLOCK_MEM, NVML_CLOCK_ID_CURRENT, &temp);
+        if (nvml_result == NVML_SUCCESS) {
+            result = (uint64_t)temp;
+        }
+        else {
+            throw Exception("NVMLDevicePool::" + std::string(__func__) +
+                            ": NVML failed to get Memory Frequency for accelerator " +
+                            std::to_string(accel_idx) + ".  Error: " + nvmlErrorString(nvml_result),
+                            GEOPM_ERROR_INVALID, __FILE__, __LINE__);
+        }
+
+        return result;
+    }
+
+    uint64_t NVMLDevicePoolImp::throttle_reasons(int accel_idx) const
+    {
+        check_accel_range(accel_idx);
+        uint64_t result = 0;
+        unsigned long long temp_ull;
```

Comment:
Defining this temp is really necessary?

Reply:
If you reorder the if to put the exception first, can get rid of `result` and return the cast temp_ull directly